source/tests/SGM/SGMGratingVelocityTests.cpp

```diff
@@ -0,0 +1,263 @@
+#include "SGMGratingVelocityTests.h"
+
+#include "beamline/AMPVControl.h"
+#include <QDebug>
+#include "actions3/AMListAction3.h"
+#include "actions3/AMActionSupport.h"
+#include "beamline/AMControlSet.h"
+#include "actions3/actions/AMWaitAction.h"
+#include "beamline/CLS/CLSMAXvMotor.h"
+#include "beamline/SGM/energy/SGMGratingTranslationStepControl.h"
+GratingVelocityTestConditions::GratingVelocityTestConditions(double encStart, double encEnd, double gratingTransTarget, double velo, double veloBase, double accel)
+{
+	encoderStart = encStart;
+	encoderEnd = encEnd;
+	gratingTranslationTarget = gratingTransTarget;
+	velocity = velo;
+	velocityBase = veloBase;
+	acceleration = accel;
+}
+
+SGMGratingVelocityTests::SGMGratingVelocityTests(QObject* parent) :
+	QObject(parent)
+{
+	gratingAngleEncoder_ =  new AMPVwStatusControl("Grating Angle",
+												   "SMTR16114I1002:enc:fbk",
+												   "SMTR16114I1002:encTarget",
+												   "SMTR16114I1002:status",
+												   "SMTR16114I1002:stop",
+												   this,
+												   5,
+												   2.0,
+												   new CLSMAXvControlStatusChecker(),
+												   1);
+
+	gratingTranslation_ = new SGMGratingTranslationStepControl(this);
+
+	gratingAngleVelocitySetpoint_ = new AMSinglePVControl("Grating Angle Velocity Setpoint",
+														  "SMTR16114I1002:velo",
+														  this,
+														  0.1);
+
+	gratingAngleVelocityFeedback_ = new AMSinglePVControl("Grating angle velocity feedback",
+														  "SMTR16114I1002:velo:fbk",
+														  this,
+														  0.1);
+
+	gratingAngleVelocityBase_ = new AMSinglePVControl("Grating angle velocity base",
+													  "SMTR16114I1002:veloBase",
+													  this,
+													  0.1);
+
+	gratingAngleAcceleration_ = new AMSinglePVControl("Grating angle acceleration",
+													  "SMTR16114I1002:accel",
+													  this,
+													  0.1);
+
+	allControls_ = new AMControlSet(this);
+	connectedOnce_ = false;
+
+
+	allControls_->addControl(gratingAngleEncoder_);
+	allControls_->addControl(gratingTranslation_);
+	allControls_->addControl(gratingAngleVelocityBase_);
+	allControls_->addControl(gratingAngleAcceleration_);
+
+	if(allControls_->isConnected()) {
+
+		onAllControlsConnected(true);
+	} else {
+
+		connect(allControls_, SIGNAL(connected(bool)), this, SLOT(onAllControlsConnected(bool)));
+	}
+
+	timer_.setInterval(500);
+	connect(&timer_, SIGNAL(timeout()), this, SLOT(onTimerTimeout()));
+
+	currentTestAction_ = 0;
+}
+
+void SGMGratingVelocityTests::testGratingAngle()
+{
+	// From 270ev (-412450) to 320eV (-348005)
+	testQueue_.push_back(GratingVelocityTestConditions(-412450, -310000, 61300, 6000, 0, 5000));
+//	testQueue_.push_back(GratingVelocityTestConditions(-412450, -313205, 61300, 3276, 0, 5000));
```

Comment:
I think I'll uncomment them all, as I commented these at the end when I was testing something that only needed one result.